sklearn/impute/_base.py

```diff
@@ -628,7 +645,15 @@ def _fit(self, X, y=None):
             The imputer mask of the original data.
 
         """
-        X = self._validate_input(X)
+        self._precomputed = False
+        if sparse.issparse(X) or isinstance(X, np.ndarray):
```

Comment:
I think that the following would be enough to detect that we deal with sparse matrix or numpy array

```python
if hasattr(X, "dtype") and X.dtype.kind == "b":
    self._precomputed = True
```